extmod/modnetwork.c

```diff
@@ -56,17 +56,48 @@ char mod_network_country_code[2] = "XX";
 #error "MICROPY_PY_NETWORK_HOSTNAME_DEFAULT must be set in mpconfigport.h or mpconfigboard.h"
 #endif
 
-char mod_network_hostname[MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN] = MICROPY_PY_NETWORK_HOSTNAME_DEFAULT;
+// We store the hostname override in a qstr so that its allocation is handled
+// by the QSTR pool and we can safely pass a pointer to it to the network
+// stack without worrying about the GC collecting the string data.
+STATIC qstr mod_network_hostname_override = MP_QSTRnull;
 
-#ifdef MICROPY_PORT_NETWORK_INTERFACES
+const char *mod_network_get_hostname(void) {
+    if (mod_network_hostname_override != MP_QSTRnull) {
+        return qstr_str(mod_network_hostname_override);
+    } else {
+        return MICROPY_PY_NETWORK_HOSTNAME_DEFAULT;
+    }
+}
+
+mp_obj_t mod_network_get_hostname_obj(void) {
+    if (mod_network_hostname_override != MP_QSTRnull) {
+        return MP_OBJ_NEW_QSTR(mod_network_hostname_override);
+    } else {
+        MP_STATIC_ASSERT((sizeof(MICROPY_PY_NETWORK_HOSTNAME_DEFAULT) - 1) <= MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN);
+        return mp_obj_new_str(MICROPY_PY_NETWORK_HOSTNAME_DEFAULT, strlen(MICROPY_PY_NETWORK_HOSTNAME_DEFAULT));
+    }
+}
+
+void mod_network_set_hostname_obj(mp_obj_t value) {
+    size_t len;
+    mp_obj_str_get_data(value, &len);
+    if (len > MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN) {
+        mp_raise_ValueError(NULL);
+    }
+    mod_network_hostname_override = mp_obj_str_get_qstr(value);
+}
 
 void mod_network_init(void) {
+    #ifdef MICROPY_PORT_NETWORK_INTERFACES
     mp_obj_list_init(&MP_STATE_PORT(mod_network_nic_list), 0);
+    #endif
 }
 
 void mod_network_deinit(void) {
+    mod_network_hostname_override = MP_QSTRnull;
```

Comment:
This is a problem... the network interface (eg WLAN) usually stays up over a soft reset, but now the hostname is being reset.

In particular a call to `netif_set_hostname()` will retain a pointer in to the qstr data, which may after a soft reset point to garbage.  And may point to a "string" that's longer than the maximum allowed value.

Reply:
Maybe we can revisit in the future. I think it is reasonable that soft reset should reset the hostname, the application code is going to be re-initialising the network interface on startup anyway and calling `network.hostname` anyway.

Anyway, in the meantime see #12549